rfd/0164-scoped-rbac.md

```diff
@@ -0,0 +1,956 @@
+dynamic mapping to SSO via connectors and on-demand assignment via access lists and access requests.
+
+Most users start with local static mapping and SSO mapping, later graduating to access requests and access lists.
+
+https://goteleport.com/docs/access-controls/sso/#creating-an-authentication-connector
+
+For SSO users, on login, Teleport checks SSO connector, and maps attributes of a user to a list of roles:
+
+```yaml
+# connector.yaml
+kind: saml
+version: v2
+metadata:
+  name: corporate
+spec:
+  attributes_to_roles:
+    - {name: "groups", value: "okta-admin", roles: ["access"]}
+     # regular expressions with capture are also supported. 
+     # the next line instructs Teleport
+     # to assign users to roles `admin-1` if his SAML "group" 
+     # attribute equals 'ssh_admin_1':
+     - { name: "group", value: "^ssh_admin_(.*)$", roles: ["admin-$1"] }
+```
+
+This creates several security and scalability issues:
+
+* Every time administrators need to change roles assignment, they have to update the resource.
+* Every time the Identity Provider changes user’s attributes, the users would have to relogin to get new roles. 
+
+For local users, administrators have to specify roles and traits in the local resource:
+
+```yaml
+kind: user
+version: v2
+metadata:
+  name: joe
+spec:
+  # roles is a list of roles assigned to this user
+  roles:
+  - admin
+  # traits are key, list of values pairs assigned to a user resource.
+  # Traits can be used in role templates as variables.
+  traits:
+    logins:
+    - joe
+    - root
+```
+
+This also creates some challenges, as administrators are forced to update local resources each time they have to assign a user new permission.
+
+### Modifications
+
+Let’s now introduce the missing pieces of the puzzle and review how new roles will simplify or deprecate legacy concepts.
+
+#### Hierarchical Resource Groups
+
+Resource groups are one such missing piece - in most cloud environments, resources are split into hierarchy, 
+for example, host `luna` is a member of a resource group `lab`, in turn a member of environment `prod`, which is in turn a member of a Teleport cluster.
+
+In Teleport, we will make cluster a default root of this hierarchy. Every computing resource by default will be a direct member of a cluster resource group.
+
+We will also restrict a resource to be a member of only one resource group at a time.  
+It’s tempting to have a resource to be assigned to multiple resource groups, but we push this out of the scope of this RFD for simplicity.
+
+In our example, the cluster administrator would define lab resource group in the following way:
+
+```yaml
+kind: resource_group
+metadata:
+  name: lab
+spec:
+  parent: prod_env
+```
+
+Administrators can assign resources to resource groups:
+
+```yaml
+kind: resource_group
+metadata:
+  name: lab
+spec:
+   parent: prod_env
+ match_kinds:
+   - node
+   - database
+   - role
+   - access_list
+   - '.*'
+ match_labels:  
+   env: prod
```

Comment:
with multiple labels here would we do a union or an intersection on matches? I bet there will be asks for the opposite of whatever we do and some combination of both and eventually full label expressions... it would be hard to do that and keep the uniqueness constraint though

Reply:
I think we can relax the constraint, let me dig into cases when we need constraints. In some specific use-cases, folks can use API to add resources.

Reply:
I want to keep it as simple as possible initially and I specified that we will use the same logic as auto discovery today. We can always make it more sophisticated if people ask.